// @mcp-use/inspector@6.0.0/dist/server/server.js
import {
  registerInspectorRoutes
} from "./chunk-TB5355KZ.js";
import "./chunk-K3ARHDEZ.js";
import {
  setServerPort
} from "./chunk-QQ37LGLJ.js";
import {
  hasNoOpenFlag,
  isPortAvailable,
  parsePortFromArgs
} from "./chunk-KD7K27XK.js";
import "./chunk-CVECQ7BJ.js";
import {
  registerStaticRoutesWithDevProxy
} from "./chunk-3UU5PBIU.js";
import "./chunk-4CQ52PTY.js";
import "./chunk-IDMHJUEM.js";
import "./chunk-PKBMQBKP.js";

// src/server/server.ts
import { serve } from "@hono/node-server";
import { Hono } from "hono";
import { cors } from "hono/cors";
import { logger } from "hono/logger";
import open from "open";
var app = new Hono();
app.use(
  "*",
  cors({
    origin: "*",
    allowHeaders: [
      "Authorization",
      "Content-Type",
      "Accept",
      "X-Target-URL",
      "X-MCP-Target",
      "Mcp-Session-Id",
      "mcp-session-id",
      "mcp-protocol-version",
      "X-Server-Id",
      "X-Requested-With",
      "X-Connection-URL"
    ],
    exposeHeaders: ["*"]
  })
);
var noisyPaths = [
  "/inspector/api/tel/",
  "/inspector/api/rpc/stream",
  "/inspector/api/rpc/log"
];
app.use(
  "/inspector/api/*",
  logger((message, ...rest) => {
    if (noisyPaths.some((p) => message.includes(p))) return;
    console.log(message, ...rest);
  })
);
registerInspectorRoutes(app);
var OAUTH_POPUP_CLOSED_HTML = `<!doctype html>
<html lang="en"><head><meta charset="utf-8"><title>Signed in</title><meta name="robots" content="noindex"><style>html,body{margin:0;height:100%;display:flex;align-items:center;justify-content:center;font-family:-apple-system,BlinkMacSystemFont,"Segoe UI",sans-serif;color:#4b5563;background:#fff}</style></head>
<body><div>Signed in. You can close this window.</div>
<script>try{if(window.opener&&!window.opener.closed)window.opener.postMessage({type:"manufact:oauth-complete"},"*")}catch(e){}try{window.close()}catch(e){}</script>
</body></html>`;
app.get(
  "/inspector/oauth-popup-closed.html",
  (c) => c.html(OAUTH_POPUP_CLOSED_HTML)
);
registerStaticRoutesWithDevProxy(app, void 0, {
  manufactChatUrl: process.env.MANUFACT_CHAT_URL
});
async function startServer() {
  try {
    const isDev = process.env.NODE_ENV === "development" || process.env.VITE_DEV === "true";
    const cliPort = parsePortFromArgs();
    let port = cliPort ?? (isDev ? 3001 : 3e3);
    const available = await isPortAvailable(port);
    if (!available) {
      if (cliPort !== null) {
        console.error(
          `\u274C Port ${port} is not available. Please stop the process using this port and try again.`
        );
        process.exit(1);
      }
      if (isDev) {
        console.error(
          `\u274C\u274C\u274C Port ${port} is not available (probably used by Vite dev server as fallback so you should stop port 3000). Please stop the process using this port and try again.`
        );
        process.exit(1);
      } else {
        const fallbackPort = 3002;
        console.warn(
          `\u26A0\uFE0F  Port ${port} is not available, trying ${fallbackPort}`
        );
        const fallbackAvailable = await isPortAvailable(fallbackPort);
        if (!fallbackAvailable) {
          console.error(
            `\u274C Neither port ${port} nor ${fallbackPort} is available. Please stop the processes using these ports and try again.`
          );
          process.exit(1);
        }
        port = fallbackPort;
      }
    }
    serve({
      fetch: app.fetch,
      port
    });
    setServerPort(port);
    if (isDev) {
      console.warn(
        `\u{1F680} MCP Inspector API server running on http://localhost:${port}`
      );
      console.warn(
        `\u{1F310} Vite dev server should be running on http://localhost:3000`
      );
    } else {
      console.warn(`\u{1F680} MCP Inspector running on http://localhost:${port}`);
    }
    if (process.env.NODE_ENV !== "production" && !hasNoOpenFlag()) {
      try {
        const url = isDev ? "http://localhost:3000" : `http://localhost:${port}`;
        await open(url);
        console.warn(`\u{1F310} Browser opened automatically`);
      } catch {
        const url = isDev ? "http://localhost:3000" : `http://localhost:${port}`;
        console.warn(`\u{1F310} Please open ${url} in your browser`);
      }
    }
    return { port, fetch: app.fetch };
  } catch (error) {
    console.error("Failed to start server:", error);
    process.exit(1);
  }
}
if (import.meta.url === `file://${process.argv[1]}`) {
  startServer();
}
var server_default = { startServer };
export {
  server_default as default
};
