// mcp-use@1.28.0/dist/src/auth/index-node.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/auth/index-node.ts
var index_node_exports = {};
__export(index_node_exports, {
  FileKVStore: () => FileKVStore,
  NodeOAuthClientProvider: () => NodeOAuthClientProvider,
  OAuthFlowError: () => OAuthFlowError,
  UnauthorizedError: () => import_auth2.UnauthorizedError,
  auth: () => import_auth2.auth
});
module.exports = __toCommonJS(index_node_exports);

// src/auth/node-provider.ts
var import_node_net = require("net");
var import_node_http = require("http");

// src/auth/file-kv-store.ts
var import_node_fs = require("fs");
var import_node_os = require("os");
var import_node_path = require("path");
var isWindows = process.platform === "win32";
var FileKVStore = class {
  static {
    __name(this, "FileKVStore");
  }
  dir;
  constructor(serverUrlHash, baseDir) {
    const root = baseDir ?? (0, import_node_path.join)((0, import_node_os.homedir)(), ".mcp-use", "oauth");
    this.dir = (0, import_node_path.join)(root, serverUrlHash);
    this.ensureDir();
  }
  ensureDir() {
    if (!(0, import_node_fs.existsSync)(this.dir)) {
      (0, import_node_fs.mkdirSync)(this.dir, { recursive: true, mode: 448 });
    }
    if (!isWindows) {
      try {
        (0, import_node_fs.chmodSync)(this.dir, 448);
      } catch {
      }
    }
  }
  /**
   * `OAuthSessionStore` keys contain colons (e.g. `mcp:auth_<hash>_tokens`)
   * which are invalid on Windows. Replace anything outside the safe set with
   * `_` so the same logical key maps to a stable filename across platforms.
   */
  sanitize(key) {
    return key.replace(/[^a-zA-Z0-9._-]/g, "_");
  }
  pathFor(key) {
    return (0, import_node_path.join)(this.dir, this.sanitize(key));
  }
  get(key) {
    const file = this.pathFor(key);
    if (!(0, import_node_fs.existsSync)(file)) return null;
    try {
      return (0, import_node_fs.readFileSync)(file, "utf-8");
    } catch {
      return null;
    }
  }
  set(key, value) {
    this.ensureDir();
    const file = this.pathFor(key);
    const tmp = `${file}.tmp.${process.pid}.${Date.now()}.${Math.random().toString(36).slice(2)}`;
    (0, import_node_fs.writeFileSync)(tmp, value, { encoding: "utf-8", mode: 384 });
    if (!isWindows) {
      try {
        (0, import_node_fs.chmodSync)(tmp, 384);
      } catch {
      }
    }
    (0, import_node_fs.renameSync)(tmp, file);
  }
  remove(key) {
    const file = this.pathFor(key);
    if ((0, import_node_fs.existsSync)(file)) {
      (0, import_node_fs.rmSync)(file, { force: true });
    }
  }
  keys() {
    if (!(0, import_node_fs.existsSync)(this.dir)) return [];
    return (0, import_node_fs.readdirSync)(this.dir).filter((name) => !name.includes(".tmp."));
  }
};

// src/auth/oauth-session-store.ts
var import_auth = require("@modelcontextprotocol/sdk/client/auth.js");

// src/utils/url-sanitize.ts
function sanitizeUrl(raw) {
  const abort = /* @__PURE__ */ __name(() => {
    throw new Error(`Invalid url to pass to open(): ${raw}`);
  }, "abort");
  let url;
  try {
    url = new URL(raw);
  } catch (_) {
    abort();
  }
  if (url.protocol !== "https:" && url.protocol !== "http:") abort();
  if (url.hostname !== encodeURIComponent(url.hostname)) abort();
  if (url.username) url.username = encodeURIComponent(url.username);
  if (url.password) url.password = encodeURIComponent(url.password);
  url.pathname = url.pathname.slice(0, 1) + encodeURIComponent(url.pathname.slice(1)).replace(/%2f/gi, "/");
  url.search = url.search.slice(0, 1) + Array.from(url.searchParams.entries()).map(sanitizeParam).join("&");
  url.hash = url.hash.slice(0, 1) + encodeURIComponent(url.hash.slice(1));
  return url.href;
}
__name(sanitizeUrl, "sanitizeUrl");
function sanitizeParam([k, v]) {
  return `${encodeURIComponent(k)}${v.length > 0 ? `=${encodeURIComponent(v)}` : ""}`;
}
__name(sanitizeParam, "sanitizeParam");

// src/auth/oauth-session-store.ts
var OAuthSessionStore = class _OAuthSessionStore {
  static {
    __name(this, "OAuthSessionStore");
  }
  serverUrl;
  storageKeyPrefix;
  serverUrlHash;
  clientName;
  clientUri;
  logoUri;
  callbackUrl;
  scope;
  store;
  _cachedAuthServerUrl = null;
  _cachedMetadata = null;
  _refreshPromise = null;
  constructor(serverUrl, options, store) {
    this.serverUrl = serverUrl;
    this.storageKeyPrefix = options.storageKeyPrefix || "mcp:auth";
    this.serverUrlHash = _OAuthSessionStore.hashString(serverUrl);
    this.clientName = options.clientName || "mcp-use";
    this.clientUri = options.clientUri || (typeof window !== "undefined" ? window.location.origin : "");
    this.logoUri = options.logoUri || "https://mcp-use.com/logo.png";
    this.callbackUrl = sanitizeUrl(
      options.callbackUrl || (typeof window !== "undefined" ? new URL("/oauth/callback", window.location.origin).toString() : "/oauth/callback")
    );
    this.scope = options.scope;
    this.store = store;
  }
  getKey(keySuffix) {
    return `${this.storageKeyPrefix}_${this.serverUrlHash}_${keySuffix}`;
  }
  static hashString(str) {
    let hash = 0;
    for (let i = 0; i < str.length; i++) {
      const char = str.charCodeAt(i);
      hash = (hash << 5) - hash + char;
      hash = hash & hash;
    }
    return Math.abs(hash).toString(16);
  }
  // --- SDK Interface Methods (delegated) ---
  get redirectUrl() {
    return this.callbackUrl;
  }
  get clientMetadata() {
    return {
      redirect_uris: [this.redirectUrl],
      token_endpoint_auth_method: "none",
      grant_types: ["authorization_code", "refresh_token"],
      response_types: ["code"],
      client_name: this.clientName,
      client_uri: this.clientUri,
      logo_uri: this.logoUri,
      ...this.scope ? { scope: this.scope } : {}
    };
  }
  async tokens() {
    const key = this.getKey("tokens");
    const data = await this.store.get(key);
    if (!data) return void 0;
    try {
      const tokens = JSON.parse(data);
      if (tokens.access_token && tokens.refresh_token) {
        try {
          const payload = JSON.parse(atob(tokens.access_token.split(".")[1]));
          if (payload.exp && Date.now() >= (payload.exp - 30) * 1e3) {
            console.log("[tokens] Access token expiring soon, refreshing...");
            const refreshed = await this._dedupedRefresh(tokens);
            if (refreshed) {
              console.log("[tokens] Refreshed successfully");
              return refreshed;
            }
          }
        } catch {
        }
      }
      return tokens;
    } catch (e) {
      console.warn(`[${this.storageKeyPrefix}] Failed to parse tokens:`, e);
      await this.store.remove(key);
      return void 0;
    }
  }
  async saveTokens(tokens) {
    await this.store.set(this.getKey("tokens"), JSON.stringify(tokens));
    await this.store.remove(this.getKey("code_verifier"));
    await this.store.remove(this.getKey("last_auth_url"));
  }
  async clientInformation() {
    const key = this.getKey("client_info");
    const data = await this.store.get(key);
    if (!data) return void 0;
    try {
      const clientInfo = JSON.parse(data);
      const storedRedirectUris = Array.isArray(clientInfo.redirect_uris) ? clientInfo.redirect_uris : [];
      const hasMatchingRedirect = storedRedirectUris.length === 0 || storedRedirectUris.includes(this.redirectUrl);
      if (!hasMatchingRedirect) {
        console.info(
          `[${this.storageKeyPrefix}] Invalidating cached OAuth client info due to redirect URI mismatch.`
        );
        await this.store.remove(key);
        await this.store.remove(this.getKey("tokens"));
        await this.store.remove(this.getKey("last_auth_url"));
        return void 0;
      }
      return clientInfo;
    } catch (e) {
      console.warn(
        `[${this.storageKeyPrefix}] Failed to parse client information:`,
        e
      );
      await this.store.remove(key);
      return void 0;
    }
  }
  async saveClientInformation(clientInformation) {
    const key = this.getKey("client_info");
    await this.store.set(key, JSON.stringify(clientInformation));
  }
  async saveCodeVerifier(codeVerifier) {
    await this.store.set(this.getKey("code_verifier"), codeVerifier);
  }
  async codeVerifier() {
    const key = this.getKey("code_verifier");
    const verifier = await this.store.get(key);
    if (!verifier) {
      throw new Error(
        `[${this.storageKeyPrefix}] Code verifier not found in storage for key ${key}. Auth flow likely corrupted or timed out.`
      );
    }
    return verifier;
  }
  async invalidateCredentials(scope) {
    switch (scope) {
      case "all":
        await this.store.remove(this.getKey("tokens"));
        await this.store.remove(this.getKey("client_info"));
        await this.store.remove(this.getKey("code_verifier"));
        await this.store.remove(this.getKey("last_auth_url"));
        break;
      case "client":
        await this.store.remove(this.getKey("client_info"));
        break;
      case "tokens":
        await this.store.remove(this.getKey("tokens"));
        break;
      case "verifier":
        await this.store.remove(this.getKey("code_verifier"));
        break;
      default:
        break;
    }
  }
  // --- Helper / non-SDK methods ---
  /**
   * Generates and persists `StoredState` for an authorization request,
   * appends the `state` query param to the URL, and persists the sanitized
   * URL to `last_auth_url` so it can be replayed on popup-blocker fallback.
   *
   * Does NOT perform any browser/proxy-specific resource rewriting — the
   * caller (platform provider) should rewrite the URL before invoking.
   *
   * @returns The sanitized authorization URL string with the `state` param appended.
   */
  async storeAuthorizationState(authorizationUrl, opts = {}) {
    const state = globalThis.crypto.randomUUID();
    const stateKey = `${this.storageKeyPrefix}:state_${state}`;
    const codeVerifier = await this.store.get(this.getKey("code_verifier"));
    const stateData = {
      serverUrlHash: this.serverUrlHash,
      expiry: Date.now() + 1e3 * 60 * 10,
      // State expires in 10 minutes
      codeVerifier: codeVerifier || void 0,
      providerOptions: {
        serverUrl: this.serverUrl,
        storageKeyPrefix: this.storageKeyPrefix,
        clientName: this.clientName,
        clientUri: this.clientUri,
        callbackUrl: this.callbackUrl,
        ...opts.extraProviderOptions ?? {}
      },
      flowType: opts.flowType,
      returnUrl: opts.returnUrl
    };
    authorizationUrl.searchParams.set("state", state);
    const sanitizedAuthUrl = sanitizeUrl(authorizationUrl.toString());
    await this.store.set(stateKey, JSON.stringify(stateData));
    await this.store.set(this.getKey("last_auth_url"), sanitizedAuthUrl);
    return sanitizedAuthUrl;
  }
  // --- Refresh logic ---
  async _refresh(tokens) {
    try {
      if (!this._cachedAuthServerUrl || !this._cachedMetadata) {
        const resourceMetadata = await (0, import_auth.discoverOAuthProtectedResourceMetadata)(
          this.serverUrl
        );
        const authServerUrl = resourceMetadata.authorization_servers?.[0];
        if (!authServerUrl) return null;
        const metadata = await (0, import_auth.discoverAuthorizationServerMetadata)(authServerUrl);
        if (!metadata) return null;
        this._cachedAuthServerUrl = authServerUrl;
        this._cachedMetadata = metadata;
      }
      const clientInfo = await this.clientInformation();
      if (!clientInfo) return null;
      const newTokens = await (0, import_auth.refreshAuthorization)(this._cachedAuthServerUrl, {
        metadata: this._cachedMetadata,
        clientInformation: clientInfo,
        refreshToken: tokens.refresh_token
      });
      await this.saveTokens(newTokens);
      return newTokens;
    } catch {
      return null;
    }
  }
  async _dedupedRefresh(tokens) {
    if (this._refreshPromise) return this._refreshPromise;
    this._refreshPromise = this._refresh(tokens);
    try {
      return await this._refreshPromise;
    } finally {
      this._refreshPromise = null;
    }
  }
  /**
   * Unconditionally exchange the persisted refresh_token for a new access
   * token. Returns the new tokens on success, or `null` if no refresh_token
   * is stored or the server rejected the grant. Shares the in-process
   * coalescing window with the lazy refresh in `tokens()`.
   */
  async forceRefresh() {
    const data = await this.store.get(this.getKey("tokens"));
    if (!data) return null;
    let stored;
    try {
      stored = JSON.parse(data);
    } catch {
      return null;
    }
    if (!stored.refresh_token) return null;
    return this._dedupedRefresh(stored);
  }
};

// src/auth/node-provider.ts
var DEFAULT_PORT = 33418;
var PORT_RANGE = 10;
var DEFAULT_AUTH_TIMEOUT_MS = 5 * 6e4;
var OAuthFlowError = class extends Error {
  static {
    __name(this, "OAuthFlowError");
  }
  code;
  description;
  constructor(code, description) {
    super(description ? `${code}: ${description}` : code);
    this.code = code;
    this.description = description;
    this.name = "OAuthFlowError";
  }
};
function createDeferred() {
  let resolve;
  let reject;
  const promise = new Promise((res, rej) => {
    resolve = res;
    reject = rej;
  });
  return { promise, resolve, reject };
}
__name(createDeferred, "createDeferred");
async function isPortFree(port) {
  return new Promise((resolve) => {
    const tester = (0, import_node_net.createServer)();
    tester.once("error", () => resolve(false));
    tester.once("listening", () => {
      tester.close(() => resolve(true));
    });
    tester.listen(port, "127.0.0.1");
  });
}
__name(isPortFree, "isPortFree");
async function reservePort(preferred, range) {
  for (let p = preferred; p < preferred + range; p++) {
    if (await isPortFree(p)) return p;
  }
  return null;
}
__name(reservePort, "reservePort");
var SUCCESS_HTML = `<!DOCTYPE html>
<html><head><meta charset="utf-8"><title>Authentication complete</title>
<style>body{font-family:system-ui,sans-serif;max-width:480px;margin:80px auto;padding:0 24px;color:#222}
h1{font-size:20px;margin:0 0 12px}p{line-height:1.5}</style></head>
<body><h1>Authentication complete</h1>
<p>You can close this tab and return to your terminal.</p></body></html>`;
function escapeHtml(s) {
  return s.replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/>/g, "&gt;").replace(/"/g, "&quot;").replace(/'/g, "&#39;");
}
__name(escapeHtml, "escapeHtml");
var FAILURE_HTML = /* @__PURE__ */ __name((err, desc) => `<!DOCTYPE html>
<html><head><meta charset="utf-8"><title>Authentication failed</title>
<style>body{font-family:system-ui,sans-serif;max-width:480px;margin:80px auto;padding:0 24px;color:#222}
h1{font-size:20px;margin:0 0 12px;color:#b00020}p{line-height:1.5}code{background:#f3f3f3;padding:2px 6px;border-radius:3px}</style></head>
<body><h1>Authentication failed</h1>
<p><code>${escapeHtml(err)}</code>${desc ? `: ${escapeHtml(desc)}` : ""}</p>
<p>You can close this tab and return to your terminal.</p></body></html>`, "FAILURE_HTML");
var NodeOAuthClientProvider = class _NodeOAuthClientProvider {
  static {
    __name(this, "NodeOAuthClientProvider");
  }
  serverUrl;
  port;
  session;
  kv;
  authTimeoutMs;
  openBrowserOverride;
  server = null;
  /** Currently in-flight deferred — used to prevent overlapping flows. */
  pending = null;
  /** Latest deferred (settled or in-flight) — what `getAuthorizationCode()` returns. */
  lastFlow = null;
  pendingTimer = null;
  constructor(serverUrl, port, session, kv, options) {
    this.serverUrl = serverUrl;
    this.port = port;
    this.session = session;
    this.kv = kv;
    this.authTimeoutMs = options.authTimeoutMs ?? DEFAULT_AUTH_TIMEOUT_MS;
    this.openBrowserOverride = options.openBrowser;
  }
  static async create(serverUrl, options = {}) {
    const serverUrlHash = OAuthSessionStore.hashString(serverUrl);
    const kv = options.kvStore ?? new FileKVStore(serverUrlHash, options.baseDir);
    const persistedPortRaw = await kv.get("port");
    const persistedPort = persistedPortRaw ? Number.parseInt(persistedPortRaw, 10) : null;
    const preferred = persistedPort && Number.isFinite(persistedPort) ? persistedPort : options.preferredPort ?? DEFAULT_PORT;
    const range = options.portRange ?? PORT_RANGE;
    let port = await reservePort(preferred, range);
    if (port === null) {
      port = await new Promise((resolve, reject) => {
        const probe = (0, import_node_net.createServer)();
        probe.once("error", reject);
        probe.once("listening", () => {
          const addr = probe.address();
          const p = typeof addr === "object" && addr ? addr.port : 0;
          probe.close(() => resolve(p));
        });
        probe.listen(0, "127.0.0.1");
      });
    }
    if (port !== persistedPort) {
      await kv.set("port", String(port));
    }
    const callbackUrl = `http://127.0.0.1:${port}/callback`;
    const session = new OAuthSessionStore(
      serverUrl,
      { ...options, callbackUrl },
      kv
    );
    return new _NodeOAuthClientProvider(serverUrl, port, session, kv, options);
  }
  // --- Identity passthroughs (parallel to BrowserOAuthClientProvider) ---
  get storageKeyPrefix() {
    return this.session.storageKeyPrefix;
  }
  get serverUrlHash() {
    return this.session.serverUrlHash;
  }
  // --- SDK Interface (delegated to OAuthSessionStore) ---
  get redirectUrl() {
    return this.session.redirectUrl;
  }
  get clientMetadata() {
    return this.session.clientMetadata;
  }
  tokens() {
    return this.session.tokens();
  }
  saveTokens(tokens) {
    return this.session.saveTokens(tokens);
  }
  clientInformation() {
    return this.session.clientInformation();
  }
  saveClientInformation(info) {
    return this.session.saveClientInformation(info);
  }
  codeVerifier() {
    return this.session.codeVerifier();
  }
  saveCodeVerifier(codeVerifier) {
    return this.session.saveCodeVerifier(codeVerifier);
  }
  invalidateCredentials(scope) {
    return this.session.invalidateCredentials(scope);
  }
  /**
   * Bind the loopback server, set up the pending-code deferred, and ask the
   * platform to open the user's browser. Does NOT await the code; the
   * orchestrator awaits via `getAuthorizationCode()`.
   */
  async redirectToAuthorization(authorizationUrl) {
    if (this.pending) {
      throw new Error(
        "NodeOAuthClientProvider: an authorization is already in progress"
      );
    }
    const sanitizedUrl = await this.session.storeAuthorizationState(
      authorizationUrl,
      { flowType: "redirect" }
    );
    await this.startLoopback();
    this.pending = createDeferred();
    this.lastFlow = this.pending;
    this.pending.promise.catch(() => {
    });
    this.pendingTimer = setTimeout(() => {
      this.rejectPending(
        new OAuthFlowError(
          "timeout",
          `No callback received within ${this.authTimeoutMs}ms`
        )
      );
    }, this.authTimeoutMs);
    const opener = this.openBrowserOverride ?? defaultOpener;
    try {
      await opener(sanitizedUrl);
    } catch (err) {
      console.error(
        `[mcp-use] Could not open browser automatically: ${err instanceof Error ? err.message : String(err)}`
      );
    }
  }
  /**
   * Resolves with the authorization code captured by the loopback callback.
   * Must be called after `redirectToAuthorization()`. Returns the same
   * promise whether the callback has fired or not — callers may subscribe
   * before or after.
   */
  getAuthorizationCode() {
    if (!this.lastFlow) {
      return Promise.reject(
        new Error(
          "NodeOAuthClientProvider.getAuthorizationCode() called before redirectToAuthorization()"
        )
      );
    }
    return this.lastFlow.promise;
  }
  /**
   * Force-refresh the access token using the persisted refresh_token.
   * Returns the new tokens on success, or null if no refresh_token / refresh failed.
   */
  forceRefresh() {
    return this.session.forceRefresh();
  }
  /**
   * Cancel an in-progress flow (timeout, SIGINT, etc.) and close the loopback.
   */
  dispose() {
    if (this.pending) {
      this.rejectPending(new OAuthFlowError("cancelled", "Flow cancelled"));
    } else {
      this.stopLoopback();
    }
  }
  /** Best-effort port for tests / status output. */
  get callbackPort() {
    return this.port;
  }
  /**
   * True if `redirectToAuthorization()` has been called and we're awaiting
   * a callback (loopback bound, browser opened). Lets orchestrators detect
   * when the SDK transport has already kicked off the flow on a 401, so they
   * can skip straight to `getAuthorizationCode()` instead of calling `auth()`
   * again (which would throw "already in progress").
   */
  get hasPendingFlow() {
    return this.pending !== null;
  }
  // --- Loopback internals ---
  async startLoopback() {
    if (this.server) return;
    const server = (0, import_node_http.createServer)((req, res) => {
      this.handleCallback(req.url ?? "/", res);
    });
    this.server = server;
    await new Promise((resolve, reject) => {
      server.once("error", reject);
      server.listen(this.port, "127.0.0.1", () => {
        server.removeListener("error", reject);
        resolve();
      });
    });
  }
  stopLoopback() {
    if (this.pendingTimer) {
      clearTimeout(this.pendingTimer);
      this.pendingTimer = null;
    }
    if (this.server) {
      this.server.close();
      this.server = null;
    }
  }
  resolvePending(code) {
    const p = this.pending;
    this.pending = null;
    this.stopLoopback();
    p?.resolve(code);
  }
  rejectPending(err) {
    const p = this.pending;
    this.pending = null;
    this.stopLoopback();
    p?.reject(err);
  }
  handleCallback(rawUrl, res) {
    const url = new URL(rawUrl, `http://127.0.0.1:${this.port}`);
    if (url.pathname !== "/callback") {
      res.statusCode = 404;
      res.end("Not Found");
      return;
    }
    const code = url.searchParams.get("code");
    const state = url.searchParams.get("state");
    const err = url.searchParams.get("error");
    const errDesc = url.searchParams.get("error_description") ?? void 0;
    if (err) {
      res.statusCode = 400;
      res.setHeader("content-type", "text/html; charset=utf-8");
      res.end(FAILURE_HTML(err, errDesc));
      this.rejectPending(new OAuthFlowError(err, errDesc));
      return;
    }
    if (!code || !state) {
      res.statusCode = 400;
      res.end("Missing code or state");
      return;
    }
    res.statusCode = 200;
    res.setHeader("content-type", "text/html; charset=utf-8");
    res.end(SUCCESS_HTML);
    this.resolvePending(code);
  }
};
async function defaultOpener(url) {
  const { spawn } = await import("child_process");
  const { platform } = await import("process");
  const cmd = platform === "darwin" ? "open" : platform === "win32" ? "cmd" : "xdg-open";
  const args = platform === "win32" ? ["/c", "start", "", url] : [url];
  await new Promise((resolve, reject) => {
    const child = spawn(cmd, args, { stdio: "ignore", detached: true });
    child.once("error", reject);
    child.once("spawn", () => {
      child.unref();
      resolve();
    });
  });
}
__name(defaultOpener, "defaultOpener");

// src/auth/index-node.ts
var import_auth2 = require("@modelcontextprotocol/sdk/client/auth.js");
