// mcp-use@1.28.0/dist/src/auth/index-node.js
import {
  OAuthSessionStore
} from "../../chunk-MMRBRXYD.js";
import {
  __name
} from "../../chunk-3GQAWCBQ.js";

// src/auth/node-provider.ts
import { createServer as createNetServer } from "net";
import { createServer as createHttpServer } from "http";

// src/auth/file-kv-store.ts
import {
  chmodSync,
  existsSync,
  mkdirSync,
  readdirSync,
  readFileSync,
  renameSync,
  rmSync,
  writeFileSync
} from "fs";
import { homedir } from "os";
import { join } from "path";
var isWindows = process.platform === "win32";
var FileKVStore = class {
  static {
    __name(this, "FileKVStore");
  }
  dir;
  constructor(serverUrlHash, baseDir) {
    const root = baseDir ?? join(homedir(), ".mcp-use", "oauth");
    this.dir = join(root, serverUrlHash);
    this.ensureDir();
  }
  ensureDir() {
    if (!existsSync(this.dir)) {
      mkdirSync(this.dir, { recursive: true, mode: 448 });
    }
    if (!isWindows) {
      try {
        chmodSync(this.dir, 448);
      } catch {
      }
    }
  }
  /**
   * `OAuthSessionStore` keys contain colons (e.g. `mcp:auth_<hash>_tokens`)
   * which are invalid on Windows. Replace anything outside the safe set with
   * `_` so the same logical key maps to a stable filename across platforms.
   */
  sanitize(key) {
    return key.replace(/[^a-zA-Z0-9._-]/g, "_");
  }
  pathFor(key) {
    return join(this.dir, this.sanitize(key));
  }
  get(key) {
    const file = this.pathFor(key);
    if (!existsSync(file)) return null;
    try {
      return readFileSync(file, "utf-8");
    } catch {
      return null;
    }
  }
  set(key, value) {
    this.ensureDir();
    const file = this.pathFor(key);
    const tmp = `${file}.tmp.${process.pid}.${Date.now()}.${Math.random().toString(36).slice(2)}`;
    writeFileSync(tmp, value, { encoding: "utf-8", mode: 384 });
    if (!isWindows) {
      try {
        chmodSync(tmp, 384);
      } catch {
      }
    }
    renameSync(tmp, file);
  }
  remove(key) {
    const file = this.pathFor(key);
    if (existsSync(file)) {
      rmSync(file, { force: true });
    }
  }
  keys() {
    if (!existsSync(this.dir)) return [];
    return readdirSync(this.dir).filter((name) => !name.includes(".tmp."));
  }
};

// src/auth/node-provider.ts
var DEFAULT_PORT = 33418;
var PORT_RANGE = 10;
var DEFAULT_AUTH_TIMEOUT_MS = 5 * 6e4;
var OAuthFlowError = class extends Error {
  static {
    __name(this, "OAuthFlowError");
  }
  code;
  description;
  constructor(code, description) {
    super(description ? `${code}: ${description}` : code);
    this.code = code;
    this.description = description;
    this.name = "OAuthFlowError";
  }
};
function createDeferred() {
  let resolve;
  let reject;
  const promise = new Promise((res, rej) => {
    resolve = res;
    reject = rej;
  });
  return { promise, resolve, reject };
}
__name(createDeferred, "createDeferred");
async function isPortFree(port) {
  return new Promise((resolve) => {
    const tester = createNetServer();
    tester.once("error", () => resolve(false));
    tester.once("listening", () => {
      tester.close(() => resolve(true));
    });
    tester.listen(port, "127.0.0.1");
  });
}
__name(isPortFree, "isPortFree");
async function reservePort(preferred, range) {
  for (let p = preferred; p < preferred + range; p++) {
    if (await isPortFree(p)) return p;
  }
  return null;
}
__name(reservePort, "reservePort");
var SUCCESS_HTML = `<!DOCTYPE html>
<html><head><meta charset="utf-8"><title>Authentication complete</title>
<style>body{font-family:system-ui,sans-serif;max-width:480px;margin:80px auto;padding:0 24px;color:#222}
h1{font-size:20px;margin:0 0 12px}p{line-height:1.5}</style></head>
<body><h1>Authentication complete</h1>
<p>You can close this tab and return to your terminal.</p></body></html>`;
function escapeHtml(s) {
  return s.replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/>/g, "&gt;").replace(/"/g, "&quot;").replace(/'/g, "&#39;");
}
__name(escapeHtml, "escapeHtml");
var FAILURE_HTML = /* @__PURE__ */ __name((err, desc) => `<!DOCTYPE html>
<html><head><meta charset="utf-8"><title>Authentication failed</title>
<style>body{font-family:system-ui,sans-serif;max-width:480px;margin:80px auto;padding:0 24px;color:#222}
h1{font-size:20px;margin:0 0 12px;color:#b00020}p{line-height:1.5}code{background:#f3f3f3;padding:2px 6px;border-radius:3px}</style></head>
<body><h1>Authentication failed</h1>
<p><code>${escapeHtml(err)}</code>${desc ? `: ${escapeHtml(desc)}` : ""}</p>
<p>You can close this tab and return to your terminal.</p></body></html>`, "FAILURE_HTML");
var NodeOAuthClientProvider = class _NodeOAuthClientProvider {
  static {
    __name(this, "NodeOAuthClientProvider");
  }
  serverUrl;
  port;
  session;
  kv;
  authTimeoutMs;
  openBrowserOverride;
  server = null;
  /** Currently in-flight deferred — used to prevent overlapping flows. */
  pending = null;
  /** Latest deferred (settled or in-flight) — what `getAuthorizationCode()` returns. */
  lastFlow = null;
  pendingTimer = null;
  constructor(serverUrl, port, session, kv, options) {
    this.serverUrl = serverUrl;
    this.port = port;
    this.session = session;
    this.kv = kv;
    this.authTimeoutMs = options.authTimeoutMs ?? DEFAULT_AUTH_TIMEOUT_MS;
    this.openBrowserOverride = options.openBrowser;
  }
  static async create(serverUrl, options = {}) {
    const serverUrlHash = OAuthSessionStore.hashString(serverUrl);
    const kv = options.kvStore ?? new FileKVStore(serverUrlHash, options.baseDir);
    const persistedPortRaw = await kv.get("port");
    const persistedPort = persistedPortRaw ? Number.parseInt(persistedPortRaw, 10) : null;
    const preferred = persistedPort && Number.isFinite(persistedPort) ? persistedPort : options.preferredPort ?? DEFAULT_PORT;
    const range = options.portRange ?? PORT_RANGE;
    let port = await reservePort(preferred, range);
    if (port === null) {
      port = await new Promise((resolve, reject) => {
        const probe = createNetServer();
        probe.once("error", reject);
        probe.once("listening", () => {
          const addr = probe.address();
          const p = typeof addr === "object" && addr ? addr.port : 0;
          probe.close(() => resolve(p));
        });
        probe.listen(0, "127.0.0.1");
      });
    }
    if (port !== persistedPort) {
      await kv.set("port", String(port));
    }
    const callbackUrl = `http://127.0.0.1:${port}/callback`;
    const session = new OAuthSessionStore(
      serverUrl,
      { ...options, callbackUrl },
      kv
    );
    return new _NodeOAuthClientProvider(serverUrl, port, session, kv, options);
  }
  // --- Identity passthroughs (parallel to BrowserOAuthClientProvider) ---
  get storageKeyPrefix() {
    return this.session.storageKeyPrefix;
  }
  get serverUrlHash() {
    return this.session.serverUrlHash;
  }
  // --- SDK Interface (delegated to OAuthSessionStore) ---
  get redirectUrl() {
    return this.session.redirectUrl;
  }
  get clientMetadata() {
    return this.session.clientMetadata;
  }
  tokens() {
    return this.session.tokens();
  }
  saveTokens(tokens) {
    return this.session.saveTokens(tokens);
  }
  clientInformation() {
    return this.session.clientInformation();
  }
  saveClientInformation(info) {
    return this.session.saveClientInformation(info);
  }
  codeVerifier() {
    return this.session.codeVerifier();
  }
  saveCodeVerifier(codeVerifier) {
    return this.session.saveCodeVerifier(codeVerifier);
  }
  invalidateCredentials(scope) {
    return this.session.invalidateCredentials(scope);
  }
  /**
   * Bind the loopback server, set up the pending-code deferred, and ask the
   * platform to open the user's browser. Does NOT await the code; the
   * orchestrator awaits via `getAuthorizationCode()`.
   */
  async redirectToAuthorization(authorizationUrl) {
    if (this.pending) {
      throw new Error(
        "NodeOAuthClientProvider: an authorization is already in progress"
      );
    }
    const sanitizedUrl = await this.session.storeAuthorizationState(
      authorizationUrl,
      { flowType: "redirect" }
    );
    await this.startLoopback();
    this.pending = createDeferred();
    this.lastFlow = this.pending;
    this.pending.promise.catch(() => {
    });
    this.pendingTimer = setTimeout(() => {
      this.rejectPending(
        new OAuthFlowError(
          "timeout",
          `No callback received within ${this.authTimeoutMs}ms`
        )
      );
    }, this.authTimeoutMs);
    const opener = this.openBrowserOverride ?? defaultOpener;
    try {
      await opener(sanitizedUrl);
    } catch (err) {
      console.error(
        `[mcp-use] Could not open browser automatically: ${err instanceof Error ? err.message : String(err)}`
      );
    }
  }
  /**
   * Resolves with the authorization code captured by the loopback callback.
   * Must be called after `redirectToAuthorization()`. Returns the same
   * promise whether the callback has fired or not — callers may subscribe
   * before or after.
   */
  getAuthorizationCode() {
    if (!this.lastFlow) {
      return Promise.reject(
        new Error(
          "NodeOAuthClientProvider.getAuthorizationCode() called before redirectToAuthorization()"
        )
      );
    }
    return this.lastFlow.promise;
  }
  /**
   * Force-refresh the access token using the persisted refresh_token.
   * Returns the new tokens on success, or null if no refresh_token / refresh failed.
   */
  forceRefresh() {
    return this.session.forceRefresh();
  }
  /**
   * Cancel an in-progress flow (timeout, SIGINT, etc.) and close the loopback.
   */
  dispose() {
    if (this.pending) {
      this.rejectPending(new OAuthFlowError("cancelled", "Flow cancelled"));
    } else {
      this.stopLoopback();
    }
  }
  /** Best-effort port for tests / status output. */
  get callbackPort() {
    return this.port;
  }
  /**
   * True if `redirectToAuthorization()` has been called and we're awaiting
   * a callback (loopback bound, browser opened). Lets orchestrators detect
   * when the SDK transport has already kicked off the flow on a 401, so they
   * can skip straight to `getAuthorizationCode()` instead of calling `auth()`
   * again (which would throw "already in progress").
   */
  get hasPendingFlow() {
    return this.pending !== null;
  }
  // --- Loopback internals ---
  async startLoopback() {
    if (this.server) return;
    const server = createHttpServer((req, res) => {
      this.handleCallback(req.url ?? "/", res);
    });
    this.server = server;
    await new Promise((resolve, reject) => {
      server.once("error", reject);
      server.listen(this.port, "127.0.0.1", () => {
        server.removeListener("error", reject);
        resolve();
      });
    });
  }
  stopLoopback() {
    if (this.pendingTimer) {
      clearTimeout(this.pendingTimer);
      this.pendingTimer = null;
    }
    if (this.server) {
      this.server.close();
      this.server = null;
    }
  }
  resolvePending(code) {
    const p = this.pending;
    this.pending = null;
    this.stopLoopback();
    p?.resolve(code);
  }
  rejectPending(err) {
    const p = this.pending;
    this.pending = null;
    this.stopLoopback();
    p?.reject(err);
  }
  handleCallback(rawUrl, res) {
    const url = new URL(rawUrl, `http://127.0.0.1:${this.port}`);
    if (url.pathname !== "/callback") {
      res.statusCode = 404;
      res.end("Not Found");
      return;
    }
    const code = url.searchParams.get("code");
    const state = url.searchParams.get("state");
    const err = url.searchParams.get("error");
    const errDesc = url.searchParams.get("error_description") ?? void 0;
    if (err) {
      res.statusCode = 400;
      res.setHeader("content-type", "text/html; charset=utf-8");
      res.end(FAILURE_HTML(err, errDesc));
      this.rejectPending(new OAuthFlowError(err, errDesc));
      return;
    }
    if (!code || !state) {
      res.statusCode = 400;
      res.end("Missing code or state");
      return;
    }
    res.statusCode = 200;
    res.setHeader("content-type", "text/html; charset=utf-8");
    res.end(SUCCESS_HTML);
    this.resolvePending(code);
  }
};
async function defaultOpener(url) {
  const { spawn } = await import("child_process");
  const { platform } = await import("process");
  const cmd = platform === "darwin" ? "open" : platform === "win32" ? "cmd" : "xdg-open";
  const args = platform === "win32" ? ["/c", "start", "", url] : [url];
  await new Promise((resolve, reject) => {
    const child = spawn(cmd, args, { stdio: "ignore", detached: true });
    child.once("error", reject);
    child.once("spawn", () => {
      child.unref();
      resolve();
    });
  });
}
__name(defaultOpener, "defaultOpener");

// src/auth/index-node.ts
import {
  auth,
  UnauthorizedError
} from "@modelcontextprotocol/sdk/client/auth.js";
export {
  FileKVStore,
  NodeOAuthClientProvider,
  OAuthFlowError,
  UnauthorizedError,
  auth
};
