// mcp-use@1.22.1-canary.1/dist/chunk-CQTMUGLH.js
import {
  __name
} from "./chunk-3GQAWCBQ.js";

// src/server/utils/proxy-client.ts
import { z } from "zod";
function jsonSchemaToZod(schema) {
  if (!schema || typeof schema !== "object") return z.any();
  if (schema.type === "object") {
    const shape = {};
    if (schema.properties) {
      for (const [key, prop] of Object.entries(schema.properties)) {
        let propSchema = jsonSchemaToZod(prop);
        if (prop.description) {
          propSchema = propSchema.describe(prop.description);
        }
        if (!schema.required?.includes(key)) {
          propSchema = propSchema.optional();
        }
        shape[key] = propSchema;
      }
    }
    return z.object(shape);
  }
  if (schema.type === "string") return z.string();
  if (schema.type === "number" || schema.type === "integer") return z.number();
  if (schema.type === "boolean") return z.boolean();
  if (schema.type === "array") {
    return schema.items ? z.array(jsonSchemaToZod(schema.items)) : z.array(z.any());
  }
  return z.any();
}
__name(jsonSchemaToZod, "jsonSchemaToZod");
async function mountSession(parentServer, childSession, namespace) {
  const prefix = namespace ? `${namespace}_` : "";
  const prefixName = /* @__PURE__ */ __name((name) => `${prefix}${name}`, "prefixName");
  try {
    const tools = await childSession.listTools();
    for (const tool of tools) {
      const toolName = prefixName(tool.name);
      parentServer.tool(
        {
          name: toolName,
          description: tool.description,
          // Convert raw JSON Schema from the child server to a Zod schema.
          // This ensures compatibility with the underlying MCP SDK which expects Zod.
          schema: jsonSchemaToZod(tool.inputSchema)
        },
        async (params, ctx) => {
          const result = await childSession.callTool(tool.name, params, {
            // Forward progress tokens if they exist
            ...ctx?._meta?.progressToken && {
              progressToken: ctx._meta.progressToken
            },
            onProgress: /* @__PURE__ */ __name(async (progress) => {
              if (ctx?.reportProgress) {
                await ctx.reportProgress(
                  progress.progress,
                  progress.total,
                  progress.message
                );
              }
            }, "onProgress")
          });
          if (result.isError) {
            throw new Error(
              result.content[0]?.type === "text" ? result.content[0].text : "Tool error"
            );
          }
          return { content: result.content, _meta: result._meta };
        }
      );
    }
  } catch (error) {
    console.warn(
      `[Proxy] Failed to mount tools for ${namespace || "unnamed"}:`,
      error
    );
  }
  try {
    const result = await childSession.listResources();
    for (const res of result.resources) {
      const resName = prefixName(res.name);
      const proxyUri = namespace ? `${namespace}://${encodeURIComponent(res.uri)}` : res.uri;
      parentServer.resource(
        {
          name: resName,
          uri: proxyUri,
          title: res.title,
          description: res.description,
          mimeType: res.mimeType
        },
        async () => {
          const originalUri = res.uri;
          const readResult = await childSession.readResource(originalUri);
          return { contents: readResult.contents };
        }
      );
    }
  } catch (error) {
    console.warn(
      `[Proxy] Failed to mount resources for ${namespace || "unnamed"}:`,
      error
    );
  }
  try {
    const result = await childSession.listPrompts();
    for (const prompt of result.prompts) {
      const promptName = prefixName(prompt.name);
      parentServer.prompt(
        {
          name: promptName,
          description: prompt.description,
          schema: z.object(
            prompt.arguments?.reduce(
              (acc, arg) => {
                let propSchema = z.string();
                if (arg.description) {
                  propSchema = propSchema.describe(arg.description);
                }
                acc[arg.name] = arg.required ? propSchema : propSchema.optional();
                return acc;
              },
              {}
            ) || {}
          )
        },
        async (params) => {
          const promptResult = await childSession.getPrompt(
            prompt.name,
            params || {}
          );
          return {
            messages: promptResult.messages,
            description: promptResult.description,
            _meta: promptResult._meta
          };
        }
      );
    }
  } catch (error) {
    console.warn(
      `[Proxy] Failed to mount prompts for ${namespace || "unnamed"}:`,
      error
    );
  }
  childSession.on("notification", (notification) => {
    switch (notification.method) {
      case "notifications/tools/list_changed":
        parentServer.nativeServer.server.sendToolListChanged().catch(console.error);
        break;
      case "notifications/resources/list_changed":
        parentServer.nativeServer.server.sendResourceListChanged().catch(console.error);
        break;
      case "notifications/prompts/list_changed":
        parentServer.nativeServer.server.sendPromptListChanged().catch(console.error);
        break;
      default:
        break;
    }
  });
}
__name(mountSession, "mountSession");

export {
  mountSession
};
