// mcp-use@1.28.0/dist/chunk-O7IWYNLB.js
import {
  BaseMCPClient,
  HttpConnector,
  normalizeClientInfo,
  resolveCallbacks
} from "./chunk-TO444N7M.js";
import {
  Tel,
  getPackageVersion
} from "./chunk-AHTMQIEU.js";
import {
  logger
} from "./chunk-QWQYAQCK.js";
import {
  __name
} from "./chunk-3GQAWCBQ.js";

// src/client/browser.ts
function trackBrowserClientInit(config) {
  const servers = Object.keys(config.mcpServers ?? {});
  Tel.getInstance().trackMCPClientInit({
    codeMode: false,
    sandbox: false,
    allCallbacks: false,
    verify: false,
    servers,
    numServers: servers.length,
    isBrowser: true
  }).catch((e) => logger.debug(`Failed to track BrowserMCPClient init: ${e}`));
}
__name(trackBrowserClientInit, "trackBrowserClientInit");
var BrowserMCPClient = class _BrowserMCPClient extends BaseMCPClient {
  static {
    __name(this, "BrowserMCPClient");
  }
  /**
   * Get the mcp-use package version.
   * Works in all environments (Node.js, browser, Cloudflare Workers, Deno, etc.)
   */
  static getPackageVersion() {
    return getPackageVersion();
  }
  constructor(config) {
    super(config);
    trackBrowserClientInit(this.config);
  }
  static fromDict(cfg) {
    return new _BrowserMCPClient(cfg);
  }
  /**
   * Create a connector from server configuration (Browser version)
   * Supports HTTP connector only
   */
  createConnectorFromConfig(serverConfig) {
    const {
      url,
      headers,
      fetch,
      authToken,
      authProvider,
      wrapTransport,
      clientOptions,
      disableSseFallback,
      preferSse,
      timeout,
      sseReadTimeout,
      gatewayUrl,
      serverId,
      reconnectionOptions
    } = serverConfig;
    if (!url) {
      throw new Error("Server URL is required");
    }
    const globalDefaults = this.config;
    const resolved = resolveCallbacks(
      serverConfig,
      globalDefaults
    );
    const clientInfo = normalizeClientInfo(
      serverConfig.clientInfo ?? this.config.clientInfo
    );
    const connectorOptions = {
      headers,
      fetch,
      authToken,
      authProvider,
      wrapTransport,
      clientOptions,
      onSampling: resolved.onSampling,
      onElicitation: resolved.onElicitation,
      onNotification: resolved.onNotification,
      disableSseFallback,
      preferSse,
      timeout,
      sseReadTimeout,
      clientInfo,
      gatewayUrl,
      serverId,
      reconnectionOptions
    };
    logger.debug(
      `[BrowserMCPClient] Connector options prepared (clientOptions: ${clientOptions ? "provided" : "none"})`
    );
    return new HttpConnector(url, connectorOptions);
  }
};

// src/utils/url-sanitize.ts
function sanitizeUrl(raw) {
  const abort = /* @__PURE__ */ __name(() => {
    throw new Error(`Invalid url to pass to open(): ${raw}`);
  }, "abort");
  let url;
  try {
    url = new URL(raw);
  } catch (_) {
    abort();
  }
  if (url.protocol !== "https:" && url.protocol !== "http:") abort();
  if (url.hostname !== encodeURIComponent(url.hostname)) abort();
  if (url.username) url.username = encodeURIComponent(url.username);
  if (url.password) url.password = encodeURIComponent(url.password);
  url.pathname = url.pathname.slice(0, 1) + encodeURIComponent(url.pathname.slice(1)).replace(/%2f/gi, "/");
  url.search = url.search.slice(0, 1) + Array.from(url.searchParams.entries()).map(sanitizeParam).join("&");
  url.hash = url.hash.slice(0, 1) + encodeURIComponent(url.hash.slice(1));
  return url.href;
}
__name(sanitizeUrl, "sanitizeUrl");
function sanitizeParam([k, v]) {
  return `${encodeURIComponent(k)}${v.length > 0 ? `=${encodeURIComponent(v)}` : ""}`;
}
__name(sanitizeParam, "sanitizeParam");

// src/auth/kv-store.ts
var LocalStorageKVStore = class {
  static {
    __name(this, "LocalStorageKVStore");
  }
  get(key) {
    return localStorage.getItem(key);
  }
  set(key, value) {
    localStorage.setItem(key, value);
  }
  remove(key) {
    localStorage.removeItem(key);
  }
  keys() {
    const out = [];
    for (let i = 0; i < localStorage.length; i++) {
      const k = localStorage.key(i);
      if (k) out.push(k);
    }
    return out;
  }
};

// src/auth/oauth-session-store.ts
import {
  discoverOAuthProtectedResourceMetadata,
  discoverAuthorizationServerMetadata,
  refreshAuthorization
} from "@modelcontextprotocol/sdk/client/auth.js";
var OAuthSessionStore = class _OAuthSessionStore {
  static {
    __name(this, "OAuthSessionStore");
  }
  serverUrl;
  storageKeyPrefix;
  serverUrlHash;
  clientName;
  clientUri;
  logoUri;
  callbackUrl;
  scope;
  store;
  _cachedAuthServerUrl = null;
  _cachedMetadata = null;
  _refreshPromise = null;
  constructor(serverUrl, options, store) {
    this.serverUrl = serverUrl;
    this.storageKeyPrefix = options.storageKeyPrefix || "mcp:auth";
    this.serverUrlHash = _OAuthSessionStore.hashString(serverUrl);
    this.clientName = options.clientName || "mcp-use";
    this.clientUri = options.clientUri || (typeof window !== "undefined" ? window.location.origin : "");
    this.logoUri = options.logoUri || "https://mcp-use.com/logo.png";
    this.callbackUrl = sanitizeUrl(
      options.callbackUrl || (typeof window !== "undefined" ? new URL("/oauth/callback", window.location.origin).toString() : "/oauth/callback")
    );
    this.scope = options.scope;
    this.store = store;
  }
  getKey(keySuffix) {
    return `${this.storageKeyPrefix}_${this.serverUrlHash}_${keySuffix}`;
  }
  static hashString(str) {
    let hash = 0;
    for (let i = 0; i < str.length; i++) {
      const char = str.charCodeAt(i);
      hash = (hash << 5) - hash + char;
      hash = hash & hash;
    }
    return Math.abs(hash).toString(16);
  }
  // --- SDK Interface Methods (delegated) ---
  get redirectUrl() {
    return this.callbackUrl;
  }
  get clientMetadata() {
    return {
      redirect_uris: [this.redirectUrl],
      token_endpoint_auth_method: "none",
      grant_types: ["authorization_code", "refresh_token"],
      response_types: ["code"],
      client_name: this.clientName,
      client_uri: this.clientUri,
      logo_uri: this.logoUri,
      ...this.scope ? { scope: this.scope } : {}
    };
  }
  async tokens() {
    const key = this.getKey("tokens");
    const data = await this.store.get(key);
    if (!data) return void 0;
    try {
      const tokens = JSON.parse(data);
      if (tokens.access_token && tokens.refresh_token) {
        try {
          const payload = JSON.parse(atob(tokens.access_token.split(".")[1]));
          if (payload.exp && Date.now() >= (payload.exp - 30) * 1e3) {
            console.log("[tokens] Access token expiring soon, refreshing...");
            const refreshed = await this._dedupedRefresh(tokens);
            if (refreshed) {
              console.log("[tokens] Refreshed successfully");
              return refreshed;
            }
          }
        } catch {
        }
      }
      return tokens;
    } catch (e) {
      console.warn(`[${this.storageKeyPrefix}] Failed to parse tokens:`, e);
      await this.store.remove(key);
      return void 0;
    }
  }
  async saveTokens(tokens) {
    await this.store.set(this.getKey("tokens"), JSON.stringify(tokens));
    await this.store.remove(this.getKey("code_verifier"));
    await this.store.remove(this.getKey("last_auth_url"));
  }
  async clientInformation() {
    const key = this.getKey("client_info");
    const data = await this.store.get(key);
    if (!data) return void 0;
    try {
      const clientInfo = JSON.parse(data);
      const storedRedirectUris = Array.isArray(clientInfo.redirect_uris) ? clientInfo.redirect_uris : [];
      const hasMatchingRedirect = storedRedirectUris.length === 0 || storedRedirectUris.includes(this.redirectUrl);
      if (!hasMatchingRedirect) {
        console.info(
          `[${this.storageKeyPrefix}] Invalidating cached OAuth client info due to redirect URI mismatch.`
        );
        await this.store.remove(key);
        await this.store.remove(this.getKey("tokens"));
        await this.store.remove(this.getKey("last_auth_url"));
        return void 0;
      }
      return clientInfo;
    } catch (e) {
      console.warn(
        `[${this.storageKeyPrefix}] Failed to parse client information:`,
        e
      );
      await this.store.remove(key);
      return void 0;
    }
  }
  async saveClientInformation(clientInformation) {
    const key = this.getKey("client_info");
    await this.store.set(key, JSON.stringify(clientInformation));
  }
  async saveCodeVerifier(codeVerifier) {
    await this.store.set(this.getKey("code_verifier"), codeVerifier);
  }
  async codeVerifier() {
    const key = this.getKey("code_verifier");
    const verifier = await this.store.get(key);
    if (!verifier) {
      throw new Error(
        `[${this.storageKeyPrefix}] Code verifier not found in storage for key ${key}. Auth flow likely corrupted or timed out.`
      );
    }
    return verifier;
  }
  async invalidateCredentials(scope) {
    switch (scope) {
      case "all":
        await this.store.remove(this.getKey("tokens"));
        await this.store.remove(this.getKey("client_info"));
        await this.store.remove(this.getKey("code_verifier"));
        await this.store.remove(this.getKey("last_auth_url"));
        break;
      case "client":
        await this.store.remove(this.getKey("client_info"));
        break;
      case "tokens":
        await this.store.remove(this.getKey("tokens"));
        break;
      case "verifier":
        await this.store.remove(this.getKey("code_verifier"));
        break;
      default:
        break;
    }
  }
  // --- Helper / non-SDK methods ---
  /**
   * Generates and persists `StoredState` for an authorization request,
   * appends the `state` query param to the URL, and persists the sanitized
   * URL to `last_auth_url` so it can be replayed on popup-blocker fallback.
   *
   * Does NOT perform any browser/proxy-specific resource rewriting — the
   * caller (platform provider) should rewrite the URL before invoking.
   *
   * @returns The sanitized authorization URL string with the `state` param appended.
   */
  async storeAuthorizationState(authorizationUrl, opts = {}) {
    const state = globalThis.crypto.randomUUID();
    const stateKey = `${this.storageKeyPrefix}:state_${state}`;
    const codeVerifier = await this.store.get(this.getKey("code_verifier"));
    const stateData = {
      serverUrlHash: this.serverUrlHash,
      expiry: Date.now() + 1e3 * 60 * 10,
      // State expires in 10 minutes
      codeVerifier: codeVerifier || void 0,
      providerOptions: {
        serverUrl: this.serverUrl,
        storageKeyPrefix: this.storageKeyPrefix,
        clientName: this.clientName,
        clientUri: this.clientUri,
        callbackUrl: this.callbackUrl,
        ...opts.extraProviderOptions ?? {}
      },
      flowType: opts.flowType,
      returnUrl: opts.returnUrl
    };
    authorizationUrl.searchParams.set("state", state);
    const sanitizedAuthUrl = sanitizeUrl(authorizationUrl.toString());
    await this.store.set(stateKey, JSON.stringify(stateData));
    await this.store.set(this.getKey("last_auth_url"), sanitizedAuthUrl);
    return sanitizedAuthUrl;
  }
  // --- Refresh logic ---
  async _refresh(tokens) {
    try {
      if (!this._cachedAuthServerUrl || !this._cachedMetadata) {
        const resourceMetadata = await discoverOAuthProtectedResourceMetadata(
          this.serverUrl
        );
        const authServerUrl = resourceMetadata.authorization_servers?.[0];
        if (!authServerUrl) return null;
        const metadata = await discoverAuthorizationServerMetadata(authServerUrl);
        if (!metadata) return null;
        this._cachedAuthServerUrl = authServerUrl;
        this._cachedMetadata = metadata;
      }
      const clientInfo = await this.clientInformation();
      if (!clientInfo) return null;
      const newTokens = await refreshAuthorization(this._cachedAuthServerUrl, {
        metadata: this._cachedMetadata,
        clientInformation: clientInfo,
        refreshToken: tokens.refresh_token
      });
      await this.saveTokens(newTokens);
      return newTokens;
    } catch {
      return null;
    }
  }
  async _dedupedRefresh(tokens) {
    if (this._refreshPromise) return this._refreshPromise;
    this._refreshPromise = this._refresh(tokens);
    try {
      return await this._refreshPromise;
    } finally {
      this._refreshPromise = null;
    }
  }
  /**
   * Unconditionally exchange the persisted refresh_token for a new access
   * token. Returns the new tokens on success, or `null` if no refresh_token
   * is stored or the server rejected the grant. Shares the in-process
   * coalescing window with the lazy refresh in `tokens()`.
   */
  async forceRefresh() {
    const data = await this.store.get(this.getKey("tokens"));
    if (!data) return null;
    let stored;
    try {
      stored = JSON.parse(data);
    } catch {
      return null;
    }
    if (!stored.refresh_token) return null;
    return this._dedupedRefresh(stored);
  }
};

// src/auth/browser-provider.ts
async function serializeBody(body) {
  if (typeof body === "string") return body;
  if (body instanceof URLSearchParams || body instanceof FormData) {
    return Object.fromEntries(body.entries());
  }
  if (body instanceof Blob) return await body.text();
  return body;
}
__name(serializeBody, "serializeBody");
var BrowserOAuthClientProvider = class {
  static {
    __name(this, "BrowserOAuthClientProvider");
  }
  serverUrl;
  staticClientInfo;
  session;
  // Browser-only state
  preventAutoAuth;
  useRedirectFlow;
  oauthProxyUrl;
  connectionUrl;
  originalFetch;
  _lastOriginalResource = null;
  onPopupWindow;
  constructor(serverUrl, options = {}) {
    this.serverUrl = serverUrl;
    this.session = new OAuthSessionStore(
      serverUrl,
      options,
      new LocalStorageKVStore()
    );
    this.preventAutoAuth = options.preventAutoAuth;
    this.useRedirectFlow = options.useRedirectFlow;
    this.oauthProxyUrl = options.oauthProxyUrl;
    this.connectionUrl = options.connectionUrl;
    this.staticClientInfo = options.staticClientInfo;
    this.onPopupWindow = options.onPopupWindow;
  }
  // --- Identity / key fields exposed for callback handling ---
  get storageKeyPrefix() {
    return this.session.storageKeyPrefix;
  }
  get serverUrlHash() {
    return this.session.serverUrlHash;
  }
  get clientName() {
    return this.session.clientName;
  }
  get clientUri() {
    return this.session.clientUri;
  }
  get logoUri() {
    return this.session.logoUri;
  }
  get callbackUrl() {
    return this.session.callbackUrl;
  }
  get scope() {
    return this.session.scope;
  }
  getKey(keySuffix) {
    return this.session.getKey(keySuffix);
  }
  /**
   * Install fetch interceptor to proxy OAuth requests through the backend
   */
  installFetchInterceptor() {
    if (!this.oauthProxyUrl) {
      console.warn(
        "[BrowserOAuthProvider] No OAuth proxy URL configured, skipping fetch interceptor installation"
      );
      return;
    }
    if (!this.originalFetch) {
      this.originalFetch = window.fetch;
    } else {
      console.warn(
        "[BrowserOAuthProvider] Fetch interceptor already installed"
      );
      return;
    }
    const oauthProxyUrl = this.oauthProxyUrl;
    const connectionUrl = this.connectionUrl;
    const serverUrl = this.serverUrl;
    const originalFetch = this.originalFetch;
    console.log(
      `[BrowserOAuthProvider] Installing fetch interceptor with proxy: ${oauthProxyUrl}`
    );
    window.fetch = async (input, init) => {
      const url = typeof input === "string" ? input : input instanceof URL ? input.toString() : input.url;
      const isOAuthRequest = url.includes("/.well-known/") || url.match(/\/(register|token|authorize)$/);
      if (!isOAuthRequest) {
        return await originalFetch(input, init);
      }
      try {
        const urlObj = new URL(url);
        const proxyUrlObj = new URL(oauthProxyUrl);
        if (urlObj.origin === proxyUrlObj.origin && (urlObj.pathname.startsWith(proxyUrlObj.pathname) || url.includes("/inspector/api/oauth"))) {
          return await originalFetch(input, init);
        }
      } catch {
      }
      try {
        const isMetadata = url.includes("/.well-known/");
        const proxyEndpoint = isMetadata ? `${oauthProxyUrl}/metadata?url=${encodeURIComponent(url)}` : `${oauthProxyUrl}/proxy`;
        console.log(
          `[OAuth Proxy] Routing ${isMetadata ? "metadata" : "request"} through: ${proxyEndpoint}`
        );
        if (isMetadata) {
          const headers = {
            ...init?.headers ? Object.fromEntries(new Headers(init.headers)) : {}
          };
          if (connectionUrl) {
            headers["X-Connection-URL"] = connectionUrl;
          }
          const proxyUrl = new URL(proxyEndpoint);
          if (serverUrl) {
            proxyUrl.searchParams.set("mcp_url", serverUrl);
          }
          const metadataResponse = await originalFetch(proxyUrl.toString(), {
            ...init,
            method: "GET",
            headers
          });
          try {
            const contentType = metadataResponse.headers.get("content-type") || "";
            if (contentType.includes("application/json")) {
              const metadataJson = await metadataResponse.clone().json();
              const originalResource = metadataJson && typeof metadataJson === "object" && typeof metadataJson._original_resource === "string" ? metadataJson._original_resource : null;
              if (originalResource) {
                this._lastOriginalResource = originalResource;
              }
            }
          } catch {
          }
          return metadataResponse;
        }
        const body = init?.body ? await serializeBody(init.body) : void 0;
        const response = await originalFetch(proxyEndpoint, {
          method: "POST",
          headers: { "Content-Type": "application/json" },
          body: JSON.stringify({
            url,
            method: init?.method || "POST",
            headers: init?.headers ? Object.fromEntries(new Headers(init.headers)) : {},
            body
          })
        });
        const data = await response.json();
        return new Response(JSON.stringify(data.body), {
          status: data.status,
          statusText: data.statusText,
          headers: new Headers(data.headers)
        });
      } catch (error) {
        console.error(
          "[OAuth Proxy] Request failed, falling back to direct fetch:",
          error
        );
        return await originalFetch(input, init);
      }
    };
  }
  /**
   * Restore original fetch after OAuth flow completes
   */
  restoreFetch() {
    if (this.originalFetch) {
      console.log("[BrowserOAuthProvider] Restoring original fetch");
      window.fetch = this.originalFetch;
      this.originalFetch = void 0;
    }
  }
  // --- SDK Interface Methods (delegated) ---
  get redirectUrl() {
    return this.session.redirectUrl;
  }
  get clientMetadata() {
    return this.session.clientMetadata;
  }
  tokens() {
    return this.session.tokens();
  }
  saveTokens(tokens) {
    return this.session.saveTokens(tokens);
  }
  async clientInformation() {
    if (this.staticClientInfo) return this.staticClientInfo;
    return this.session.clientInformation();
  }
  async saveClientInformation(clientInformation) {
    if (this.staticClientInfo) return;
    return this.session.saveClientInformation(clientInformation);
  }
  codeVerifier() {
    return this.session.codeVerifier();
  }
  saveCodeVerifier(codeVerifier) {
    return this.session.saveCodeVerifier(codeVerifier);
  }
  invalidateCredentials(scope) {
    return this.session.invalidateCredentials(scope);
  }
  /**
   * Generates and persists `StoredState` for an authorization request,
   * applies browser-only resource rewriting, and returns the sanitized URL
   * with the `state` param appended. Does NOT open a popup or redirect —
   * use `redirectToAuthorization` for that.
   */
  async prepareAuthorizationUrl(authorizationUrl) {
    this.rewriteResourceIfLocalProxy(authorizationUrl);
    return this.session.storeAuthorizationState(authorizationUrl, {
      extraProviderOptions: {
        oauthProxyUrl: this.oauthProxyUrl,
        connectionUrl: this.connectionUrl,
        ...this.staticClientInfo ? { staticClientInfo: this.staticClientInfo } : {},
        ...this.scope ? { scope: this.scope } : {}
      },
      flowType: this.useRedirectFlow ? "redirect" : "popup",
      returnUrl: typeof window !== "undefined" ? window.location.href : void 0
    });
  }
  /**
   * Redirects the user agent to the authorization URL, storing necessary state.
   * @param authorizationUrl The fully constructed authorization URL from the SDK.
   */
  async redirectToAuthorization(authorizationUrl) {
    const sanitizedAuthUrl = await this.prepareAuthorizationUrl(authorizationUrl);
    if (this.preventAutoAuth) {
      console.info(
        `[${this.storageKeyPrefix}] Auto-auth prevented. Authorization URL stored for manual trigger.`
      );
      return;
    }
    if (this.useRedirectFlow) {
      console.info(
        `[${this.storageKeyPrefix}] Redirecting to authorization URL (full-page redirect).`
      );
      window.location.href = sanitizedAuthUrl;
      return;
    }
    const popupFeatures = "width=600,height=700,resizable=yes,scrollbars=yes,status=yes";
    try {
      const popup = window.open(
        sanitizedAuthUrl,
        `mcp_auth_${this.serverUrlHash}`,
        popupFeatures
      );
      if (this.onPopupWindow) {
        this.onPopupWindow(sanitizedAuthUrl, popupFeatures, popup);
      }
      if (!popup || popup.closed || typeof popup.closed === "undefined") {
        console.warn(
          `[${this.storageKeyPrefix}] Popup likely blocked by browser. Manual navigation might be required using the stored URL.`
        );
      } else {
        popup.focus();
        console.info(
          `[${this.storageKeyPrefix}] Redirecting to authorization URL in popup.`
        );
      }
    } catch (e) {
      console.error(
        `[${this.storageKeyPrefix}] Error opening popup window:`,
        e
      );
    }
  }
  // --- Browser-only helpers ---
  /**
   * If the SDK-built authorization URL has a `resource` parameter pointing
   * at the local inspector proxy (rather than the real MCP server), rewrite
   * it to the original resource so the OAuth server's allowlist matches.
   */
  rewriteResourceIfLocalProxy(url) {
    const originalResourceParam = url.searchParams.get("resource");
    const looksLikeLocalProxyResource = Boolean(
      originalResourceParam && (originalResourceParam.includes("/inspector/api/proxy") || originalResourceParam.includes("/api/proxy") || originalResourceParam.includes("localhost:3000"))
    );
    const matchesConnectionUrl = Boolean(
      originalResourceParam && this.connectionUrl && originalResourceParam === this.connectionUrl
    );
    const shouldRewriteResource = Boolean(
      originalResourceParam && (this._lastOriginalResource || this.serverUrl) && (matchesConnectionUrl || looksLikeLocalProxyResource)
    );
    const rewriteTargetResource = this._lastOriginalResource || this.serverUrl;
    if (shouldRewriteResource && rewriteTargetResource) {
      url.searchParams.set("resource", rewriteTargetResource);
    }
  }
  /**
   * Retrieves the last URL passed to `redirectToAuthorization`. Useful for manual fallback.
   */
  getLastAttemptedAuthUrl() {
    const storedUrl = localStorage.getItem(this.getKey("last_auth_url"));
    if (!storedUrl) return null;
    return sanitizeUrl(storedUrl);
  }
  clearStorage() {
    const prefixPattern = `${this.storageKeyPrefix}_${this.serverUrlHash}_`;
    const statePattern = `${this.storageKeyPrefix}:state_`;
    const keysToRemove = [];
    let count = 0;
    for (let i = 0; i < localStorage.length; i++) {
      const key = localStorage.key(i);
      if (!key) continue;
      if (key.startsWith(prefixPattern)) {
        keysToRemove.push(key);
      } else if (key.startsWith(statePattern)) {
        try {
          const item = localStorage.getItem(key);
          if (item) {
            const state = JSON.parse(item);
            if (state.serverUrlHash === this.serverUrlHash) {
              keysToRemove.push(key);
            }
          }
        } catch (e) {
          console.warn(
            `[${this.storageKeyPrefix}] Error parsing state key ${key} during clearStorage:`,
            e
          );
        }
      }
    }
    const uniqueKeysToRemove = [...new Set(keysToRemove)];
    uniqueKeysToRemove.forEach((key) => {
      localStorage.removeItem(key);
      count++;
    });
    return count;
  }
};

// src/auth/callback.ts
import { auth } from "@modelcontextprotocol/sdk/client/auth.js";
async function onMcpAuthorization() {
  const queryParams = new URLSearchParams(window.location.search);
  const code = queryParams.get("code");
  const state = queryParams.get("state");
  const error = queryParams.get("error");
  const errorDescription = queryParams.get("error_description");
  const logPrefix = "[mcp-callback]";
  console.log(`${logPrefix} Handling callback...`, {
    code,
    state,
    error,
    errorDescription
  });
  let provider = null;
  let storedStateData = null;
  let stateKey = null;
  try {
    if (!state) {
      throw new Error(
        "State parameter not found or invalid in callback query parameters."
      );
    }
    console.log(`[mcp-callback] Looking for state: ${state}`);
    console.log(
      `[mcp-callback] All localStorage keys:`,
      Object.keys(localStorage)
    );
    const defaultStateKey = `mcp:auth:state_${state}`;
    if (localStorage.getItem(defaultStateKey)) {
      stateKey = defaultStateKey;
      console.log(
        `[mcp-callback] Found state with default key: ${defaultStateKey}`
      );
    } else {
      const stateKeySuffix = `:state_${state}`;
      for (let i = 0; i < localStorage.length; i++) {
        const key = localStorage.key(i);
        if (key && key.endsWith(stateKeySuffix)) {
          stateKey = key;
          console.log(`[mcp-callback] Found state with dynamic key: ${key}`);
          break;
        }
      }
    }
    if (!stateKey) {
      const stateKeys = Object.keys(localStorage).filter(
        (k) => k.includes("state")
      );
      console.log(`[mcp-callback] State keys in storage:`, stateKeys);
      throw new Error(
        `Invalid or expired state parameter "${state}". No matching state found in storage.`
      );
    }
    const storedStateJSON = localStorage.getItem(stateKey);
    if (!storedStateJSON) {
      throw new Error(
        `Invalid or expired state parameter "${state}". No matching state found in storage.`
      );
    }
    try {
      storedStateData = JSON.parse(storedStateJSON);
    } catch (e) {
      throw new Error("Failed to parse stored OAuth state.");
    }
    if (!storedStateData.expiry || storedStateData.expiry < Date.now()) {
      localStorage.removeItem(stateKey);
      throw new Error(
        "OAuth state has expired. Please try initiating authentication again."
      );
    }
    if (error) {
      console.log(
        `${logPrefix} OAuth error received: ${error} - ${errorDescription || "No description"}`
      );
      const isRedirectFlow = storedStateData.flowType === "redirect";
      const hasOpener = window.opener && !window.opener.closed;
      const redirectWithError = /* @__PURE__ */ __name((target) => {
        console.log(`${logPrefix} Returning to: ${target}`);
        localStorage.removeItem(stateKey);
        const url = new URL(target);
        url.searchParams.set("auth_error", error);
        if (errorDescription) {
          url.searchParams.set("auth_error_description", errorDescription);
        }
        window.location.href = url.toString();
      }, "redirectWithError");
      if (storedStateData.returnUrl && (isRedirectFlow || !hasOpener)) {
        redirectWithError(storedStateData.returnUrl);
        return;
      }
      if (hasOpener) {
        window.opener.postMessage(
          {
            type: "mcp_auth_callback",
            success: false,
            error: `${error}${errorDescription ? `: ${errorDescription}` : ""}`
          },
          window.location.origin
        );
        localStorage.removeItem(stateKey);
        window.close();
        return;
      }
      throw new Error(
        `OAuth error: ${error}${errorDescription ? ` - ${errorDescription}` : ""}`
      );
    }
    if (!code) {
      throw new Error(
        "Authorization code not found in callback query parameters."
      );
    }
    if (!storedStateData.providerOptions) {
      throw new Error("Stored state is missing required provider options.");
    }
    const { serverUrl, ...providerOptions } = storedStateData.providerOptions;
    const rawConnectionUrl = providerOptions.connectionUrl;
    const isHostedGatewayConnection = Boolean(
      rawConnectionUrl && /run\.mcp-use\.com|mcp-use\.run/.test(rawConnectionUrl)
    );
    const callbackConnectionUrl = isHostedGatewayConnection ? rawConnectionUrl : void 0;
    let oauthProxyUrl = providerOptions.oauthProxyUrl;
    const connectionUrl = callbackConnectionUrl;
    if (!oauthProxyUrl && connectionUrl) {
      try {
        const callbackUrl = new URL(window.location.href);
        if (callbackUrl.pathname.includes("/oauth/callback")) {
          let basePath = callbackUrl.pathname.replace(
            /\/oauth\/callback.*$/,
            ""
          );
          if (!basePath || basePath === "") {
            const isGatewayConnection = /run\.mcp-use\.com|mcp-use\.run/.test(
              connectionUrl
            );
            if (isGatewayConnection) {
              console.log(
                `${logPrefix} Gateway connection detected; skipping default /inspector OAuth proxy inference`
              );
            } else {
              basePath = "/inspector";
              console.log(
                `${logPrefix} Callback at root /oauth/callback, using /inspector as base path for OAuth proxy`
              );
            }
          }
          if (basePath) {
            oauthProxyUrl = `${callbackUrl.origin}${basePath}/api/oauth`;
          }
          console.log(
            `${logPrefix} Inferred OAuth proxy URL from callback: ${oauthProxyUrl}`
          );
        }
      } catch (e) {
        console.warn(`${logPrefix} Could not infer OAuth proxy URL:`, e);
      }
    }
    console.log(
      `${logPrefix} Re-instantiating provider for server: ${serverUrl}`
    );
    provider = new BrowserOAuthClientProvider(serverUrl, {
      ...providerOptions,
      oauthProxyUrl,
      connectionUrl
    });
    const stateVerifierKey = `${providerOptions.storageKeyPrefix}_${storedStateData.serverUrlHash}_code_verifier`;
    const verifierForState = storedStateData.codeVerifier || localStorage.getItem(stateVerifierKey);
    if (verifierForState) {
      localStorage.setItem(stateVerifierKey, verifierForState);
      localStorage.setItem(provider.getKey("code_verifier"), verifierForState);
    } else {
      throw new Error(
        `[${providerOptions.storageKeyPrefix}] Code verifier missing for OAuth state ${state}. Please restart authentication.`
      );
    }
    if (oauthProxyUrl) {
      console.log(
        `${logPrefix} Installing fetch interceptor for token exchange (proxy: ${oauthProxyUrl})`
      );
      provider.installFetchInterceptor();
    }
    console.log(`${logPrefix} Calling SDK auth() to exchange code...`);
    const sdkServerUrl = isHostedGatewayConnection && connectionUrl ? connectionUrl : serverUrl;
    console.log(
      `${logPrefix} Using SDK serverUrl: ${sdkServerUrl} (connectionUrl: ${connectionUrl || "none"})`
    );
    const authResult = await auth(provider, {
      serverUrl: sdkServerUrl,
      authorizationCode: code
    });
    if (authResult === "AUTHORIZED") {
      console.log(`${logPrefix} Authorization successful via SDK auth().`);
      const isRedirectFlow = storedStateData.flowType === "redirect";
      if (isRedirectFlow && storedStateData.returnUrl) {
        console.log(
          `${logPrefix} Redirect flow complete. Returning to: ${storedStateData.returnUrl}`
        );
        localStorage.removeItem(stateKey);
        window.location.href = storedStateData.returnUrl;
      } else if (window.opener && !window.opener.closed) {
        console.log(`${logPrefix} Popup flow complete. Notifying opener...`);
        window.opener.postMessage(
          { type: "mcp_auth_callback", success: true },
          window.location.origin
        );
        localStorage.removeItem(stateKey);
        window.close();
      } else if (storedStateData.returnUrl) {
        console.log(
          `${logPrefix} Popup flow without opener. Returning to: ${storedStateData.returnUrl}`
        );
        localStorage.removeItem(stateKey);
        window.location.href = storedStateData.returnUrl;
      } else {
        console.warn(
          `${logPrefix} No opener window or return URL detected. Redirecting to root.`
        );
        localStorage.removeItem(stateKey);
        const pathParts = window.location.pathname.split("/").filter(Boolean);
        const basePath = pathParts.length > 0 && pathParts[pathParts.length - 1] === "callback" ? "/" + pathParts.slice(0, -2).join("/") : "/";
        window.location.href = basePath || "/";
      }
    } else {
      console.warn(
        `${logPrefix} SDK auth() returned unexpected status: ${authResult}`
      );
      throw new Error(
        `Unexpected result from authentication library: ${authResult}`
      );
    }
  } catch (err) {
    console.error(`${logPrefix} Error during OAuth callback handling:`, err);
    const errorMessage = err instanceof Error ? err.message : String(err);
    if (window.opener && !window.opener.closed) {
      window.opener.postMessage(
        { type: "mcp_auth_callback", success: false, error: errorMessage },
        window.location.origin
      );
    }
    try {
      document.body.innerHTML = "";
      const container = document.createElement("div");
      container.style.fontFamily = "sans-serif";
      container.style.padding = "20px";
      const heading = document.createElement("h1");
      heading.textContent = "Authentication Error";
      container.appendChild(heading);
      const errorPara = document.createElement("p");
      errorPara.style.color = "red";
      errorPara.style.backgroundColor = "#ffebeb";
      errorPara.style.border = "1px solid red";
      errorPara.style.padding = "10px";
      errorPara.style.borderRadius = "4px";
      errorPara.textContent = errorMessage;
      container.appendChild(errorPara);
      const closePara = document.createElement("p");
      closePara.textContent = "You can close this window or ";
      const closeLink = document.createElement("a");
      closeLink.href = "#";
      closeLink.textContent = "click here to close";
      closeLink.onclick = (e) => {
        e.preventDefault();
        window.close();
        return false;
      };
      closePara.appendChild(closeLink);
      closePara.appendChild(document.createTextNode("."));
      container.appendChild(closePara);
      if (err instanceof Error && err.stack) {
        const stackPre = document.createElement("pre");
        stackPre.style.fontSize = "0.8em";
        stackPre.style.color = "#555";
        stackPre.style.marginTop = "20px";
        stackPre.style.whiteSpace = "pre-wrap";
        stackPre.textContent = err.stack;
        container.appendChild(stackPre);
      }
      document.body.appendChild(container);
    } catch (displayError) {
      console.error(
        `${logPrefix} Could not display error in callback window:`,
        displayError
      );
    }
    if (stateKey) {
      localStorage.removeItem(stateKey);
    }
    if (provider) {
      localStorage.removeItem(provider.getKey("last_auth_url"));
    }
  }
}
__name(onMcpAuthorization, "onMcpAuthorization");

export {
  BrowserMCPClient,
  sanitizeUrl,
  BrowserOAuthClientProvider,
  onMcpAuthorization
};
