// @mcp-use/inspector@0.25.1-canary.0/dist/server/chunk-QQ37LGLJ.js
// src/server/tunnel.ts
import { spawn } from "child_process";
import { readFileSync } from "fs";
import { join } from "path";
var TUNNEL_URL_REGEX = /https?:\/\/([a-z0-9-]+\.[a-z0-9.-]+)/i;
var SETUP_TIMEOUT_MS = 3e4;
var serverPort = null;
var activeTunnel = null;
function setServerPort(port) {
  serverPort = port;
}
function getTunnelStatus() {
  if (!activeTunnel) {
    return { url: null, subdomain: null };
  }
  return { url: activeTunnel.url, subdomain: activeTunnel.subdomain };
}
function readSubdomainFromManifest() {
  const projectPath = process.env.MCP_USE_PROJECT_PATH || process.cwd();
  try {
    const raw = readFileSync(
      join(projectPath, "dist", "mcp-use.json"),
      "utf-8"
    );
    const manifest = JSON.parse(raw);
    return manifest?.tunnel?.subdomain || void 0;
  } catch {
    return void 0;
  }
}
async function startTunnel(subdomain) {
  if (activeTunnel) {
    return { url: activeTunnel.url, subdomain: activeTunnel.subdomain };
  }
  if (!serverPort) {
    throw new Error("Server port not set. Cannot start tunnel.");
  }
  const resolvedSubdomain = subdomain ?? readSubdomainFromManifest();
  const port = serverPort;
  return new Promise((resolve, reject) => {
    console.log(`[Tunnel] Starting tunnel for port ${port}...`);
    const tunnelArgs = ["--yes", "@mcp-use/tunnel", String(port)];
    if (resolvedSubdomain) {
      tunnelArgs.push("--subdomain", resolvedSubdomain);
      console.log(`[Tunnel] Reusing subdomain: ${resolvedSubdomain}`);
    }
    const proc = spawn("npx", tunnelArgs, {
      stdio: ["ignore", "pipe", "pipe"],
      shell: process.platform === "win32"
    });
    let resolved = false;
    const setupTimeout = setTimeout(() => {
      if (!resolved) {
        resolved = true;
        try {
          proc.kill("SIGTERM");
        } catch {
        }
        reject(new Error("Tunnel setup timed out after 30s"));
      }
    }, SETUP_TIMEOUT_MS);
    proc.stdout?.on("data", (data) => {
      const text = data.toString();
      const isShutdownMessage = text.includes("Shutting down") || text.includes("\u{1F6D1}");
      if (!isShutdownMessage) {
        process.stdout.write(`[Tunnel] ${text}`);
      }
      const urlMatch = text.match(TUNNEL_URL_REGEX);
      if (urlMatch && !resolved) {
        const url = urlMatch[0];
        const fullDomain = urlMatch[1];
        const subdomainMatch = fullDomain.match(/^([a-z0-9-]+)\./i);
        const extractedSubdomain = subdomainMatch ? subdomainMatch[1] : fullDomain.split(".")[0];
        resolved = true;
        clearTimeout(setupTimeout);
        activeTunnel = { url, subdomain: extractedSubdomain, process: proc };
        proc.on("exit", () => {
          if (activeTunnel?.process === proc) {
            console.log("[Tunnel] Process exited, clearing state");
            activeTunnel = null;
          }
        });
        console.log(`[Tunnel] Established: ${url}`);
        resolve({ url, subdomain: extractedSubdomain });
      }
    });
    proc.stderr?.on("data", (data) => {
      const text = data.toString();
      if (!text.includes("INFO") && !text.includes("bore_cli") && !text.includes("Shutting down")) {
        process.stderr.write(`[Tunnel] ${text}`);
      }
    });
    proc.on("error", (error) => {
      if (!resolved) {
        resolved = true;
        clearTimeout(setupTimeout);
        reject(new Error(`Failed to start tunnel: ${error.message}`));
      }
    });
    proc.on("exit", (code) => {
      if (!resolved && code !== 0) {
        resolved = true;
        clearTimeout(setupTimeout);
        reject(new Error(`Tunnel process exited with code ${code}`));
      }
    });
  });
}
function stopTunnel() {
  if (!activeTunnel) {
    return false;
  }
  try {
    activeTunnel.process.kill("SIGTERM");
  } catch {
  }
  activeTunnel = null;
  console.log("[Tunnel] Stopped");
  return true;
}

export {
  setServerPort,
  getTunnelStatus,
  startTunnel,
  stopTunnel
};
