// mcp-use@1.28.0/dist/chunk-MMRBRXYD.js
import {
  __name
} from "./chunk-3GQAWCBQ.js";

// src/utils/url-sanitize.ts
function sanitizeUrl(raw) {
  const abort = /* @__PURE__ */ __name(() => {
    throw new Error(`Invalid url to pass to open(): ${raw}`);
  }, "abort");
  let url;
  try {
    url = new URL(raw);
  } catch (_) {
    abort();
  }
  if (url.protocol !== "https:" && url.protocol !== "http:") abort();
  if (url.hostname !== encodeURIComponent(url.hostname)) abort();
  if (url.username) url.username = encodeURIComponent(url.username);
  if (url.password) url.password = encodeURIComponent(url.password);
  url.pathname = url.pathname.slice(0, 1) + encodeURIComponent(url.pathname.slice(1)).replace(/%2f/gi, "/");
  url.search = url.search.slice(0, 1) + Array.from(url.searchParams.entries()).map(sanitizeParam).join("&");
  url.hash = url.hash.slice(0, 1) + encodeURIComponent(url.hash.slice(1));
  return url.href;
}
__name(sanitizeUrl, "sanitizeUrl");
function sanitizeParam([k, v]) {
  return `${encodeURIComponent(k)}${v.length > 0 ? `=${encodeURIComponent(v)}` : ""}`;
}
__name(sanitizeParam, "sanitizeParam");

// src/auth/oauth-session-store.ts
import {
  discoverOAuthProtectedResourceMetadata,
  discoverAuthorizationServerMetadata,
  refreshAuthorization
} from "@modelcontextprotocol/sdk/client/auth.js";
var OAuthSessionStore = class _OAuthSessionStore {
  static {
    __name(this, "OAuthSessionStore");
  }
  serverUrl;
  storageKeyPrefix;
  serverUrlHash;
  clientName;
  clientUri;
  logoUri;
  callbackUrl;
  scope;
  store;
  _cachedAuthServerUrl = null;
  _cachedMetadata = null;
  _refreshPromise = null;
  constructor(serverUrl, options, store) {
    this.serverUrl = serverUrl;
    this.storageKeyPrefix = options.storageKeyPrefix || "mcp:auth";
    this.serverUrlHash = _OAuthSessionStore.hashString(serverUrl);
    this.clientName = options.clientName || "mcp-use";
    this.clientUri = options.clientUri || (typeof window !== "undefined" ? window.location.origin : "");
    this.logoUri = options.logoUri || "https://mcp-use.com/logo.png";
    this.callbackUrl = sanitizeUrl(
      options.callbackUrl || (typeof window !== "undefined" ? new URL("/oauth/callback", window.location.origin).toString() : "/oauth/callback")
    );
    this.scope = options.scope;
    this.store = store;
  }
  getKey(keySuffix) {
    return `${this.storageKeyPrefix}_${this.serverUrlHash}_${keySuffix}`;
  }
  static hashString(str) {
    let hash = 0;
    for (let i = 0; i < str.length; i++) {
      const char = str.charCodeAt(i);
      hash = (hash << 5) - hash + char;
      hash = hash & hash;
    }
    return Math.abs(hash).toString(16);
  }
  // --- SDK Interface Methods (delegated) ---
  get redirectUrl() {
    return this.callbackUrl;
  }
  get clientMetadata() {
    return {
      redirect_uris: [this.redirectUrl],
      token_endpoint_auth_method: "none",
      grant_types: ["authorization_code", "refresh_token"],
      response_types: ["code"],
      client_name: this.clientName,
      client_uri: this.clientUri,
      logo_uri: this.logoUri,
      ...this.scope ? { scope: this.scope } : {}
    };
  }
  async tokens() {
    const key = this.getKey("tokens");
    const data = await this.store.get(key);
    if (!data) return void 0;
    try {
      const tokens = JSON.parse(data);
      if (tokens.access_token && tokens.refresh_token) {
        try {
          const payload = JSON.parse(atob(tokens.access_token.split(".")[1]));
          if (payload.exp && Date.now() >= (payload.exp - 30) * 1e3) {
            console.log("[tokens] Access token expiring soon, refreshing...");
            const refreshed = await this._dedupedRefresh(tokens);
            if (refreshed) {
              console.log("[tokens] Refreshed successfully");
              return refreshed;
            }
          }
        } catch {
        }
      }
      return tokens;
    } catch (e) {
      console.warn(`[${this.storageKeyPrefix}] Failed to parse tokens:`, e);
      await this.store.remove(key);
      return void 0;
    }
  }
  async saveTokens(tokens) {
    await this.store.set(this.getKey("tokens"), JSON.stringify(tokens));
    await this.store.remove(this.getKey("code_verifier"));
    await this.store.remove(this.getKey("last_auth_url"));
  }
  async clientInformation() {
    const key = this.getKey("client_info");
    const data = await this.store.get(key);
    if (!data) return void 0;
    try {
      const clientInfo = JSON.parse(data);
      const storedRedirectUris = Array.isArray(clientInfo.redirect_uris) ? clientInfo.redirect_uris : [];
      const hasMatchingRedirect = storedRedirectUris.length === 0 || storedRedirectUris.includes(this.redirectUrl);
      if (!hasMatchingRedirect) {
        console.info(
          `[${this.storageKeyPrefix}] Invalidating cached OAuth client info due to redirect URI mismatch.`
        );
        await this.store.remove(key);
        await this.store.remove(this.getKey("tokens"));
        await this.store.remove(this.getKey("last_auth_url"));
        return void 0;
      }
      return clientInfo;
    } catch (e) {
      console.warn(
        `[${this.storageKeyPrefix}] Failed to parse client information:`,
        e
      );
      await this.store.remove(key);
      return void 0;
    }
  }
  async saveClientInformation(clientInformation) {
    const key = this.getKey("client_info");
    await this.store.set(key, JSON.stringify(clientInformation));
  }
  async saveCodeVerifier(codeVerifier) {
    await this.store.set(this.getKey("code_verifier"), codeVerifier);
  }
  async codeVerifier() {
    const key = this.getKey("code_verifier");
    const verifier = await this.store.get(key);
    if (!verifier) {
      throw new Error(
        `[${this.storageKeyPrefix}] Code verifier not found in storage for key ${key}. Auth flow likely corrupted or timed out.`
      );
    }
    return verifier;
  }
  async invalidateCredentials(scope) {
    switch (scope) {
      case "all":
        await this.store.remove(this.getKey("tokens"));
        await this.store.remove(this.getKey("client_info"));
        await this.store.remove(this.getKey("code_verifier"));
        await this.store.remove(this.getKey("last_auth_url"));
        break;
      case "client":
        await this.store.remove(this.getKey("client_info"));
        break;
      case "tokens":
        await this.store.remove(this.getKey("tokens"));
        break;
      case "verifier":
        await this.store.remove(this.getKey("code_verifier"));
        break;
      default:
        break;
    }
  }
  // --- Helper / non-SDK methods ---
  /**
   * Generates and persists `StoredState` for an authorization request,
   * appends the `state` query param to the URL, and persists the sanitized
   * URL to `last_auth_url` so it can be replayed on popup-blocker fallback.
   *
   * Does NOT perform any browser/proxy-specific resource rewriting — the
   * caller (platform provider) should rewrite the URL before invoking.
   *
   * @returns The sanitized authorization URL string with the `state` param appended.
   */
  async storeAuthorizationState(authorizationUrl, opts = {}) {
    const state = globalThis.crypto.randomUUID();
    const stateKey = `${this.storageKeyPrefix}:state_${state}`;
    const codeVerifier = await this.store.get(this.getKey("code_verifier"));
    const stateData = {
      serverUrlHash: this.serverUrlHash,
      expiry: Date.now() + 1e3 * 60 * 10,
      // State expires in 10 minutes
      codeVerifier: codeVerifier || void 0,
      providerOptions: {
        serverUrl: this.serverUrl,
        storageKeyPrefix: this.storageKeyPrefix,
        clientName: this.clientName,
        clientUri: this.clientUri,
        callbackUrl: this.callbackUrl,
        ...opts.extraProviderOptions ?? {}
      },
      flowType: opts.flowType,
      returnUrl: opts.returnUrl
    };
    authorizationUrl.searchParams.set("state", state);
    const sanitizedAuthUrl = sanitizeUrl(authorizationUrl.toString());
    await this.store.set(stateKey, JSON.stringify(stateData));
    await this.store.set(this.getKey("last_auth_url"), sanitizedAuthUrl);
    return sanitizedAuthUrl;
  }
  // --- Refresh logic ---
  async _refresh(tokens) {
    try {
      if (!this._cachedAuthServerUrl || !this._cachedMetadata) {
        const resourceMetadata = await discoverOAuthProtectedResourceMetadata(
          this.serverUrl
        );
        const authServerUrl = resourceMetadata.authorization_servers?.[0];
        if (!authServerUrl) return null;
        const metadata = await discoverAuthorizationServerMetadata(authServerUrl);
        if (!metadata) return null;
        this._cachedAuthServerUrl = authServerUrl;
        this._cachedMetadata = metadata;
      }
      const clientInfo = await this.clientInformation();
      if (!clientInfo) return null;
      const newTokens = await refreshAuthorization(this._cachedAuthServerUrl, {
        metadata: this._cachedMetadata,
        clientInformation: clientInfo,
        refreshToken: tokens.refresh_token
      });
      await this.saveTokens(newTokens);
      return newTokens;
    } catch {
      return null;
    }
  }
  async _dedupedRefresh(tokens) {
    if (this._refreshPromise) return this._refreshPromise;
    this._refreshPromise = this._refresh(tokens);
    try {
      return await this._refreshPromise;
    } finally {
      this._refreshPromise = null;
    }
  }
  /**
   * Unconditionally exchange the persisted refresh_token for a new access
   * token. Returns the new tokens on success, or `null` if no refresh_token
   * is stored or the server rejected the grant. Shares the in-process
   * coalescing window with the lazy refresh in `tokens()`.
   */
  async forceRefresh() {
    const data = await this.store.get(this.getKey("tokens"));
    if (!data) return null;
    let stored;
    try {
      stored = JSON.parse(data);
    } catch {
      return null;
    }
    if (!stored.refresh_token) return null;
    return this._dedupedRefresh(stored);
  }
};

export {
  sanitizeUrl,
  OAuthSessionStore
};
